benchmarks/distributed/rpc/parameter_server/trainer/hooks.py

```diff
@@ -0,0 +1,96 @@
+from utils import process_bucket_with_remote_server
+
+import torch
+import torch.distributed as c10d
+
+
+def allreduce_hook(state, bucket):
+    r"""
+    A ddp communication hook that uses the process_group allreduce implementation.
+    Args:
+        state (object): maintains state during the training process
+        bucket (GradBucket): gradient bucket
+    """
+    cref = state.cref
+    tensor = bucket.get_tensor()
+    tensors = [tensor / state.process_group.size()]
+    key = state.get_key(bucket.get_index())
+    if tensor.is_sparse:
+        tensor = tensor.coalesce()
+    tensor_type = "sparse" if tensor.is_sparse else "dense"
+    cref.record_start("hook_future_metric", key, f"{cref.backend}_{tensor_type}_allreduce")
+    fut = state.process_group.allreduce(tensors).get_future()
+
+    def callback(fut):
+        cref.record_end("hook_future_metric", key)
+        return fut.wait()
+
+    return fut.then(callback)
+
+
+def hybrid_hook(state, bucket):
+    r"""
+    A ddp communication hook that uses two process groups.
+    Args:
+        state (object): maintains state during the training process
+        bucket (GradBucket): gradient bucket
+    """
+    cref = state.cref
+    tensor = bucket.get_tensor()
+    key = state.get_key(bucket.get_index())
+
+    if tensor.is_sparse:
+        cref.record_start("hook_c10d_metric", key, "gloo_sparse_allreduce")
+        tensor = tensor.coalesce()
+        tensor = tensor / state.process_group.size()
+        c10d.all_reduce(tensor, op=c10d.ReduceOp.SUM)
```

Comment:
(no need to change anything) curious, any reason that we choose to initialize the default process group with gloo, and create a non-default nccl one in the state, instead of creating two non-default ones?

Reply:
No reason, just a different implementation. Only thing is the default process group async returns a work handle, not a torch future.